include/random_walks/billiard_shake_and_bake_walk.hpp

```diff
@@ -0,0 +1,229 @@
+    BilliardShakeAndBakeWalk()        : param(0, false) {}
+
+    struct parameters {
+        parameters(double L, bool set) : m_L(L), set_L(set) {}
+        double m_L;  bool set_L;
+    } param;
+
+    template <typename Polytope, typename RandomNumberGenerator>
+    struct Walk
+    {
+        using Point   = typename Polytope::PointType;
+        using NT      = typename Point::FT;
+        using VT      = typename Polytope::VT;
+        using MT      = typename Polytope::MT;
+        using DenseMT = Eigen::Matrix<NT,Eigen::Dynamic,Eigen::Dynamic>;
+        static constexpr bool SPARSE = std::is_same_v<MT, Eigen::SparseMatrix<NT, Eigen::RowMajor>>;
+
+        struct update_parameters {
+            int  facet_prev{-1};
+            NT   inner_vi_ak{NT(0)};
+            bool hit_ball{false};
+            NT   ball_inner_norm{NT(1)};
+        } params_;
+
+        static constexpr NT kDefaultEps = NT(1e-10);
+
+        template <typename GenericPolytope>
+        Walk(GenericPolytope &P,
+             const Point &boundary_pt,
+             int          facet_idx,
+             RandomNumberGenerator &rng,
+             NT           eps = kDefaultEps)
+          : P_(P), epsilon_(eps)
+        {
+            if(!P.is_normalized()) {
+                P.normalize();
+            }
+            _Len = compute_diameter<GenericPolytope>::template compute<NT>(P);
+            if constexpr (SPARSE)
+                _AA = DenseMT(P.get_mat() * P.get_mat().transpose());
+            else
+                _AA.noalias() = P.get_mat() * P.get_mat().transpose();
+            initialize(boundary_pt, facet_idx, rng);
+        }
+
+        template <typename GenericPolytope>
+        Walk(GenericPolytope &P,
+             const Point &boundary_pt,
+             int          facet_idx,
+             RandomNumberGenerator &rng,
+             NT           eps,
+             NT           manual_L)
+          : P_(P), epsilon_(eps)
+        {
+            if (!P_.is_normalized()) P_.normalize();
+            if constexpr (SPARSE)
+                _AA = DenseMT(P.get_mat() * P.get_mat().transpose());
+            else
+                _AA.noalias() = P.get_mat() * P.get_mat().transpose();
+            _Len = manual_L;
+            initialize(boundary_pt, facet_idx, rng);
+        }
+
+        NT get_epsilon() const noexcept { return epsilon_; }
+
+        void apply(unsigned walk_len, RandomNumberGenerator& rng)
+        {
+            for (unsigned step = 0; step < walk_len; ++step) {
+                params_.facet_prev = -1;
+                _lambda_prev      = NT(0);
+                lambda_hit_       = NT(0);
+
+                Point v = get_direction(rng);
+                NT    T = -std::log(rng.sample_urdist()) * _Len;  
+
+                Point p0 = p_;         
+                int   reflections = 0;
+
+                auto [first_lambda, first_facet] = P_.line_positive_intersect_skip(p_, v, Ar_, Av_, _lambda_prev,_AA, params_,params_.facet_prev);
```

Comment:
The first reflection of each step should be done via: `std::pair<NT, int> pbpair = P.line_first_positive_intersect(_p, _v, _lambdas, _Av, _update_parameters);`. Please check: https://github.com/GeomScale/volesti/blob/develop/include/random_walks/uniform_accelerated_billiard_walk.hpp